bot/slack/slack_bot.py

```diff
@@ -0,0 +1,161 @@
+import random
+import sys
+import os
+import time
+
+from slackclient import SlackClient
+
+from bot_plugin_set import BotPluginSet
+
+from mozdef_util.utilities.logger import logger
+
+
+greetings = [
+    "mozdef bot in da house",
+    "mozdef here..what's up",
+    "mozdef has joined the room..no one panic",
+    "mozdef bot here..nice to see everyone"
+]
+
+
+def formatAlert(alert_dict):
+    summary = alert_dict['summary']
+    if 'category' in alert_dict.keys():
+        summary = "_{0}_: {1}".format(alert_dict['category'], summary)
+    return summary
+
+
+class SlackBot():
+    def __init__(self, api_key, channels, bot_name, async_func):
+        self.slack_client = SlackClient(api_key)
+        self.channels = channels
+        self.bot_name = bot_name
+        self.async_func = async_func
+        self.load_commands()
+
+    def load_commands(self):
+        plugin_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), 'commands'))
+        self.plugin_set = BotPluginSet(plugin_dir)
+
+    def run(self):
+        if self.slack_client.rtm_connect():
+            logger.info("Bot connected to slack")
+            self.post_welcome_message(random.choice(greetings))
+            self.async_func(self)
+            self.listen_for_messages()
+        else:
+            logger.error("Unable to connect to slack")
+            sys.exit(1)
+
+    def handle_command(self, message_text):
+        response = ""
+        message_tokens = message_text.split()
+        command = message_tokens[0]
+        parameters = message_tokens[1:len(message_tokens)]
+
+        if command == '!help':
+            response = "\nHelp is on it's way...try these:\n"
+            for plugin in self.plugin_set.enabled_plugins:
+                response += "\n{0} -- {1}".format(
+                    plugin['command_name'],
+                    plugin['help_text']
+                )
+        else:
+            # Loop through plugins and send message to it
+            plugin_command_found = False
+            for plugin in self.plugin_set.enabled_plugins:
```

Comment:
Rather than looping through plugins to find the one we're looking for, we should store all 'registered' plugins in a dictionary contained by the `SlackBot` so that lookup can be done in constant time and more explicitly. We're using the [Command Pattern](https://en.wikipedia.org/wiki/Command_pattern) anyway, so we might as well do it properly.

Reply:
Implemented in https://github.com/mozilla/MozDef/pull/1018/commits/30346b03e5adc484378fe4953bb59b8ca8bf3cb8